sdk/src/types/block/output/feature/metadata.rs

```diff
@@ -82,6 +103,76 @@ impl StorageScore for MetadataFeature {}
 
 impl WorkScore for MetadataFeature {}
 
+impl Packable for MetadataFeature {
+    type UnpackError = Error;
+    type UnpackVisitor = ();
+
+    fn pack<P: Packer>(&self, packer: &mut P) -> Result<(), P::Error> {
+        (self.0.len() as u16).pack(packer)?;
+
+        for (k, v) in self.0.iter() {
+            k.pack(packer)?;
+            v.pack(packer)?;
+        }
+
+        Ok(())
+    }
+
+    fn unpack<U: Unpacker, const VERIFY: bool>(
+        unpacker: &mut U,
+        visitor: &Self::UnpackVisitor,
+    ) -> Result<Self, UnpackError<Self::UnpackError, U::Error>> {
+        let len = u16::unpack::<_, VERIFY>(unpacker, &()).coerce()?;
+
+        let mut map = BTreeMap::<
+            BoxedSlicePrefix<u8, MetadataFeatureKeyLength>,
+            BoxedSlicePrefix<u8, MetadataFeatureValueLength>,
+        >::new();
+
+        for _ in 0..len {
+            let key = BoxedSlicePrefix::<u8, MetadataFeatureKeyLength>::unpack::<_, VERIFY>(unpacker, visitor)
+                .map_err(|_| Error::InvalidMetadataFeature("invalid key".to_string()))
+                .map_err(UnpackError::Packable)?;
+
+            if let Some((last, _)) = map.last_key_value() {
+                match last.cmp(&key) {
+                    core::cmp::Ordering::Equal => {
+                        return Err(UnpackError::Packable(Error::InvalidMetadataFeature(
+                            "duplicated key".to_string(),
+                        )));
+                    }
+                    core::cmp::Ordering::Greater => {
+                        return Err(UnpackError::Packable(Error::InvalidMetadataFeature(
+                            "unordered map".to_string(),
+                        )));
+                    }
+                    core::cmp::Ordering::Less => (),
+                }
+            }
+
+            let value = BoxedSlicePrefix::<u8, MetadataFeatureValueLength>::unpack::<_, VERIFY>(unpacker, visitor)
+                .map_err(|_| Error::InvalidMetadataFeature("invalid value".to_string()))
+                .map_err(UnpackError::Packable)?;
+
+            map.insert(key, value);
+        }
+
+        let r: BTreeMapPrefix<
+            BoxedSlicePrefix<u8, MetadataFeatureKeyLength>,
+            BoxedSlicePrefix<u8, MetadataFeatureValueLength>,
+            MetadataFeatureLength,
+        > = map
+            .try_into()
+            .map_err(|_| Error::InvalidMetadataFeature("invalid metadata feature".to_string()))
+            .map_err(UnpackError::Packable)?;
+
+        verify_keys_packable::<true>(&r).map_err(UnpackError::Packable)?;
+        verify_length_packable::<true>(&r).map_err(UnpackError::Packable)?;
```

Comment:
Can't this be done before creating the structure, using the bytes?

Reply:
Not sure, on the Unpacker trait I don't see a method to get the length of the bytes, do you know a way?

Reply:
Looks like the only way currently is how it's done for the block:

```rust
let start_opt = unpacker.read_bytes();

...

if VERIFY {
    let block_len = if let (Some(start), Some(end)) = (start_opt, unpacker.read_bytes()) {
        end - start
    } else {
        block.packed_len()
    };
    if block_len > Self::LENGTH_MAX {
        return Err(UnpackError::Packable(Error::InvalidBlockLength(block_len)));
    }
}
```

Reply:
yeah tbh

Reply:
Just derived packable now again and moved it to the verify method where we can't do this as far as I understand

Reply:
yeah I really don't like that solution, since it means we call the pack impl on every unpack for no real reason

Reply:
So you prefer manual impl with CounterUnpacker?

Reply:
yes

Reply:
the manual impl can be extremely simple, as you can call unpack on the `BTreeSetPrefix` type instead of doing all the work manually.

Reply:
Hm I fear this is actually not possible
https://github.com/iotaledger/iota-sdk/blob/2b44b886538b3c5dbbaf55a8ace38e8f44273bfe/sdk/src/types/block/core.rs#L169 creates the CounterUnpacker in another method which accepts the bytes and only then calls the unpack() method.
Inside of the `unpack()` method there seems to be no way to create it (we need `slice: &'a [u8]` for the SliceUnpacker and if we would have that, we could also just use its length already

Reply:
We should be able to do this:

```rust
fn unpack<U: Unpacker, const VERIFY: bool>(
        unpacker: &mut U,
        visitor: &Self::UnpackVisitor,
    ) -> Result<Self, UnpackError<Self::UnpackError, U::Error>> {
  let mut unpacker = CounterUnpacker::new(unpacker);
  ...
}
```

since the `CounterUnpacker` just takes any `U: Unpacker` and `Unpacker` is implemented for `&mut U`

Reply:
Right thanks this works